[Ryan Tan Zheng Ri] iP

ryantzr1 opened the pull request:
# FRIDAY :rocket:
FRIDAY is your cool chatbot assistant. It's 
+ cool
+ fun 
+ easy to use

## Features
1. Task Management
2. Todo Management
3. Deadline Management

All you need to do is,
- [ ] download it from [here](https://github.com/ryantzr1/ip/releases/download/A-Jar/friday.jar)
- [ ] double click it 
-  [ ] run `java -jar friday.jar`

If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) {
        Friday friday = new Friday("data/tasks.txt");
        friday.start();
    }
```

> Remember: Friday is **fun**!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

ryantzr1 committed:
Complete Duke Level-0 task: Rename, Greet, Exit

ryantzr1 committed:
Level 1: Add Echo function to echo command entered by user

ryantzr1 committed:
Level 1: Update echo function

ryantzr1 committed:
Level 2: Add the ability to store text entered by user

ryantzr1 committed:
Level 3: Add the ability to mark tasks as done and not done

ryantzr1 committed:
Level 4: Add support for tracking Todos, Deadlines and Events

ryantzr1 committed:
Semi-automate the testing

ryantzr1 committed:
Level 5: Teach the bot to deal with errors

ryantzr1 committed:
Add support for deleting tasks from the list

ryantzr1 committed:
Add gitignore file

ryantzr1 committed:
Add javadocs to Duke application

ryantzr1 committed:
Update EXPECTED.TXT for automated UI testing

ryantzr1 committed:
Refactor app: Rename chatbot and convert static method to instance-based

ryantzr1 committed:
Add Save Task Functionality

ryantzr1 committed:
Merge branch 'branch-Level-7'

ryantzr1 committed:
Add Date handling method

ryantzr1 committed:
Merge branch 'branch-Level-8'

ryantzr1 committed:
Add more class for better OOP

ryantzr1 committed:
Merge branch 'A-MoreOOP'

ryantzr1 committed:
Add files into friday package

ryantzr1 committed:
Merge branch 'A-packages'

ryantzr1 committed:
Merge branch 'add-gradle-support'

ryantzr1 committed:
Add JUnit tests

ryantzr1 committed:
Merge branch 'add-JUnit'

ryantzr1 committed:
Add JAR file

ryantzr1 committed:
Merge branch 'A-Jar'

ryantzr1 committed:
Add Javadocs

ryantzr1 committed:
Merge branch 'branch-A-JavaDoc'

This is to add Javadocs for all public methods

ryantzr1 committed:
Tweak code to comply with coding standard

ryantzr1 committed:
Merge branch 'branch-A-CodingStandard'

ryantzr1 committed:
Add search function by keyword

ryantzr1 committed:
Merge branch 'branch-Level-9'

This is to add a find by keyword function.

ryantzr1 committed:
Add checkstyle plugin

ryantzr1 committed:
Add checkstyle config files

ryantzr1 committed:
Merge branch 'A-CheckStyle'

This is to add checkstyle functionality to our app.

georgetayqy commented on src/main/java/friday/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Formats the deadline string
```
Just a small nit regarding Javadocs:

```suggestion
     * Formats the deadline string.
```

georgetayqy commented on src/main/java/friday/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+     * @throws DateTimeParseException if the provided deadline cannot be parsed.
+     */
+    private String toDate() {
+        //Solution below inspired by https://www.baeldung.com/java-datetimeformatter
```
I like that you credit where you took inspiration from within your code!

georgetayqy commented on src/main/java/friday/Friday.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public static void main(String[] args) {
+        Friday friday = new Friday("data/tasks.txt");
```
Should this be declared above as a `private static final String` constant?

```suggestion
        Friday friday = new Friday(Friday.DATA_FILE_PATH);
```

georgetayqy commented on src/main/java/friday/FridayException.java:
```diff
@@ -0,0 +1,10 @@
+public class FridayException extends Exception {
+    public FridayException(String message) {
+        super(message);
+    }
```
Should the constructor have some Javadoc explaining what the parameters mean?

```suggestion
/**
 * Constructs an instance of {@code FridayException}.
 * 
 * @param message Exception message containing details of the error
 */
public FridayException(String message) {
    super(message);
}
```

georgetayqy commented on src/main/java/friday/Parser.java:
```diff
@@ -0,0 +1,107 @@
+     * Processes user inputs and interacts with the TaskList to execute user commands.
+     * Continues to run until the user inputs the "bye" command.
+     */
+    public void echo(Scanner input, TaskList taskList, Ui ui, Storage storage) {
```
Should this method be named like `runFriday` or `exectureFriday`, to better describe what the method does?

spatuly replied:
Maybe can consider handling each case in a different method as this method is getting rather long (more than 30 LoC)? It could also help with avoiding deep nesting.

georgetayqy commented on src/main/java/friday/Parser.java:
```diff
@@ -0,0 +1,107 @@
+                    taskList.add(todo);
+                    ui.taskListMessage(taskList);
+                    saveTasks(taskList.toString(), storage);
+
```
Should this extra line be removed?

georgetayqy commented on src/main/java/friday/Storage.java:
```diff
@@ -0,0 +1,41 @@
+ * Represents a storage system to save and view tasks.
+ */
+public class Storage {
+
+    private File storage;
+    private String filePath;
+    /**
```
Should there be no extra lines between the class declaration and the class attribute declarations?

```suggestion
    private File storage;
    private String filePath;
    
    /**
```

georgetayqy commented on src/main/java/friday/Task.java:
```diff
@@ -0,0 +1,51 @@
+public class Task {
+    protected String name;
+    protected boolean isDone;
+
```
Should the constructor have its own Javadoc?

```suggestion
/**
 * Constructs a new instance of {@code Task}.
 *
 * @param name The name of the task
 */
```

georgetayqy commented on src/main/java/friday/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Returns a status icon based on whether the task is done.
+     * @return A string representing the status icon.
```
Should the Javadocs have space between the comments and the @ tags?

```suggestion
     * Returns a status icon based on whether the task is done.
     *
     * @return A string representing the status icon.
```

georgetayqy commented on src/main/java/friday/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Mark the task as done.
```
Should the Javadoc for this method start with `Marks`?

georgetayqy commented on src/main/java/friday/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Mark the task as not done.
```
Should the Javadoc for this method start with Marks?

georgetayqy commented on src/main/java/friday/Task.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    /**
+     * Checks if keyword is contained in task description.
+     * @param keyWord The keyword to check for.
```
Should the Javadocs have space between the comments and the @ tags?

```suggestion
     * Checks if keyword is contained in task description.
     *
     * @param keyWord The keyword to check for.
```

georgetayqy commented on src/main/java/friday/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    /**
+     * Mark a specific task in the list as done.
+     * @param taskNumber The index of the task to be marked as done.
```
Should the Javadoc start with `Marks` instead? Also, should there be an extra line between the first line of text and the second line? I have noticed this issue in some of the other chunks of code.

georgetayqy commented on src/main/java/friday/Ui.java:
```diff
@@ -0,0 +1,95 @@
+package friday;
+public class Ui {
+
```
Should these public methods have their own Javadocs to explain the usage of the methods? I am also seeing this issue in some places within your code.

georgetayqy commented on src/test/java/friday/DeadlineTest.java:
```diff
@@ -0,0 +1,30 @@
+     * Tests that valid date formats are correctly parsed and represented in the Deadline class.
+     */
+    @Test
+    public void testValidDateFormats() {
```
Should this test method be named like `testDeadline_validDate_success`?

georgetayqy commented on src/test/java/friday/DeadlineTest.java:
```diff
@@ -0,0 +1,30 @@
+     * Tests that an invalid date format throws the expected DateTimeParseException with a relevant error message.
+     */
+    @Test
+    public void testInvalidDateFormat() {
```
Should this test method be named like `testDeadline_invalidDate_exceptionThrown`? I am seeing this issue on naming for the different test methods in other parts of the code.

georgetayqy reviewed (commented):
LGTM. Just a few nits to iron out, but overall, good job on the PR!

ryantzr1 committed:
Add GUI to the chatbot

Add JavaFX to implement GUI.

There is now a user-friendly way
to talk to Friday.

ryantzr1 committed:
Merge branch 'Level-10'

There are major changes as GUI is added.

DialogBox and main GUI window support
provided.

spatuly commented on src/main/java/friday/Parser.java:
```diff
@@ -0,0 +1,107 @@
+        while (input.hasNextLine()) {
+            try {
+                String userInput = input.nextLine();
+                if (userInput.equals("bye")) {
```
Should a switch statement be used instead of if-else? Might be easier to handle / improve readability.

spatuly commented on src/main/java/friday/TaskList.java:
```diff
@@ -0,0 +1,76 @@
+    public void delete(int taskNumber) {
+        Task task = taskList.get(taskNumber);
+        taskList.remove(taskNumber);
+        System.out.println("Noted. I've removed this task:\n" + task);
```
Could consider moving these println statements to Ui.java? I noticed several of them in this class.

spatuly commented on src/main/java/friday/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+                deadline = LocalDate.parse(deadlineStr.trim(), formatter);
+                return deadline.format(DateTimeFormatter.ofPattern("MMM d yyyy"));
+            } catch (DateTimeParseException e) {
+            }
```
Should probably try to avoid empty catch blocks? If not maybe can leave a comment explaining why it was left empty.

spatuly commented on src/main/java/friday/Deadline.java:
```diff
@@ -0,0 +1,51 @@
+            } catch (DateTimeParseException e) {
+            }
+         }
+        throw new DateTimeParseException("Unsupported date format: " + deadlineStr, deadlineStr, 0);
```
Any reason why DateTimeParseException was caught and immediately thrown again?

spatuly commented on src/main/java/friday/Parser.java:
```diff
@@ -0,0 +1,107 @@
+                    saveTasks(taskList.toString(), storage);
+                } else {
+                    ui.showBadInputError();
+                }
```
Generally for these if-else statements, maybe there could be empty lines separating some groups of related statements? Not too important but it would improve readability.

spatuly commented on src/main/java/friday/Parser.java:
```diff
@@ -0,0 +1,107 @@
+                    ui.addTask(event);
+                    taskList.add(event);
+                    ui.taskListMessage(taskList);
+                    saveTasks(taskList.toString(), storage);
```
Perhaps switch the order of printing messages and the action so it flows more logically? E.g. add the task to the taskList before "added: ..." is printed.

spatuly reviewed (commented):
Code quality is generally fine. The only significant issue is with the echo method in Parser.java, which was mentioned in the comments along with a few more trivial problems. Consider creating some more methods so echo can become... shorter. Good job otherwise :)

ryantzr1 committed:
Refactor to integrate parser

This is to make the code more OOP and remove
unnecessary code blocks.

ryantzr1 committed:
Refactor TaskList

ryantzr1 committed:
Add Java assert

Shift from using JUnit Assertions.

Introduce Java's `assert` for in-code sanity checks.

Rational: enhance early detection of unexpected behaviours.

ryantzr1 committed:
Update gradle for assertions

ryantzr1 committed:
Improve Code Quality

ryantzr1 committed:
Merge pull request #2 from ryantzr1/A-CodeQuality

Improve Code Quality

ryantzr1 committed:
Merge pull request #1 from ryantzr1/A-Assertions

Use Java's assert feature

ryantzr1 committed:
Add Github Actions

ryantzr1 committed:
Merge branch 'A-CI'

ryantzr1 committed:
Add Note Taking Functionality

Users were only able to track tasks.

Users are now able to manage their notes.

Rational:
Creating notes is important so you can track key points
in your meetings, conversations and daily life.

ryantzr1 committed:
Merge pull request #3 from ryantzr1/D-Notes

Add Note Taking Functionality

ryantzr1 committed:
Fix checkstyle issues

ryantzr1 committed:
Merge branch 'master' of https://github.com/ryantzr1/ip

ryantzr1 committed:
Fix checkstyle issues

ryantzr1 committed:
Refactor Code to Fix Deadline Fn

There was an issue with toString method in Task class.

Slight modification in Task class renaming of variables
was done.

ryantzr1 committed:
Merge branch 'fixDeadline'

ryantzr1 committed:
Update Github Actions Workflow

ryantzr1 committed:
Update Coding Style for Tests

ryantzr1 committed:
Merge branch 'updateCI'

ryantzr1 committed:
Test CI with simple commit

ryantzr1 committed:
Fix minor checkstyle error

ryantzr1 committed:
Merge pull request #5 from ryantzr1/testCI

Test CI with simple commit

ryantzr1 committed:
Organise files into structured packages

ryantzr1 committed:
Fix checkstyle errors

ryantzr1 committed:
Update test files

ryantzr1 committed:
Enhance OOP for exceptions and code cleanup

ryantzr1 committed:
Merge branch 'refactor-exceptions-oop'

Refactoring code to follow best coding practices.

Enhancing OOP of Exceptions by using polymorphism.

ryantzr1 committed:
Enhance storage handling

Refactored Storage feature to ensure better modularity.
Addressed issue where task list not stored on computer.

ryantzr1 committed:
Update User Guide and add product screenshot
